### ISSUE TITLE
apt-get update, install fails on fresh node (kvm local hyperviser)

### ISSUE THREAD

**Issue by ubuntu-server-builder** [canonical/cloud-init#2252]
This bug was originally filed in Launchpad as [LP: #966309](https://bugs.launchpad.net/cloud-init/+bug/966309)
<details>
<summary>Launchpad details</summary>
<pre>
affected_projects = []
assignee = None
assignee_name = None
date_closed = 2012-08-16T13:48:30.323472+00:00
date_created = 2012-03-27T15:19:10.732293+00:00
date_fix_committed = None
date_fix_released = None
id = 966309
importance = low
is_complete = True
lp_url = https://bugs.launchpad.net/cloud-init/+bug/966309
milestone = None
owner = rjeschmi
owner_name = Rob Schmidt
private = False
status = wont_fix
submitter = rjeschmi
submitter_name = Rob Schmidt
tags = []
duplicates = []
</pre>
</details>

_Launchpad user **Rob Schmidt(rjeschmi)** wrote on 2012-03-27T15:19:10.732293+00:00_

I have tried both oneric and precise cloud images for qemu and keep getting an error in the apt-get install (also got it in the update)

<pre>

Mar 27 15:09:56 havarti [CLOUDINIT] cc_apt_update_upgrade.py[WARNING]: Failed to install packages: ['puppet'] 
Mar 27 15:09:56 havarti [CLOUDINIT] cc_apt_update_upgrade.py[DEBUG]: Traceback (most recent call last):#012  File "/usr/lib/python2.7/dist-packages/cloudinit/CloudConfig/cc_apt_update_upgrade.py", line 79, in handle#012    cc.install_packages(pkglist)#012  File "/usr/lib/python2.7/dist-packages/cloudinit/CloudConfig/__init__.py", line 250, in install_packages#012    apt_get("install",pkglist)#012  File "/usr/lib/python2.7/dist-packages/cloudinit/CloudConfig/__init__.py", line 243, in apt_get#012    subprocess.check_call(cmd,env=e)#012  File "/usr/lib/python2.7/subprocess.py", line 511, in check_call#012    raise CalledProcessError(retcode, cmd)#012CalledProcessError: Command '['apt-get', '--option', 'Dpkg::Options::=--force-confold', '--assume-yes', 'install', 'puppet']' returned non-zero exit status 100
Mar 27 15:09:56 havarti [CLOUDINIT] __init__.py[WARNING]: Traceback (most recent call last):#012  File "/usr/lib/python2.7/dist-packages/cloudinit/CloudConfig/__init__.py", line 108, in run_cc_modules#012    cc.handle(name, run_args, freq=freq)#012  File "/usr/lib/python2.7/dist-packages/cloudinit/CloudConfig/__init__.py", line 72, in handle#012    [ name, self.cfg, self.cloud, cloudinit.log, args ])#012  File "/usr/lib/python2.7/dist-packages/cloudinit/__init__.py", line 309, in sem_and_run#012    func(*args)#012  File "/usr/lib/python2.7/dist-packages/cloudinit/CloudConfig/cc_apt_update_upgrade.py", line 86, in handle#012    raise errors[0]#012CalledProcessError: Command '['apt-get', '--option', 'Dpkg::Options::=--force-confold', '--assume-yes', 'install', 'puppet']' returned non-zero exit status 100
Mar 27 15:09:56 havarti [CLOUDINIT] __init__.py[ERROR]: config handling of apt-update-upgrade, None, [] failed
Mar 27 15:09:56 havarti [CLOUDINIT] __init__.py[DEBUG]: handling timezone with freq=None and args=[]
Mar 27 15:09:56 havarti [CLOUDINIT] __init__.py[DEBUG]: handling puppet with freq=None and args=[]

</pre>

userdata looks like this:

<pre>
packages:
  - puppet

puppet:
  conf:
    agent:
      server: "puppet.cihm"
</pre>

and I'm making a transport iso like this:

./make-iso --hostname havarti ovf-env.xml.tmpl user-data

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Rob Schmidt(rjeschmi)** wrote on 2012-03-27T15:35:26.228561+00:00_

more tests:

removed the packages: puppet

booted fine and cloud-init.log was clean

then ran:

root@havarti:/var/log# apt-get --option Dpkg::Options::=--force-confold --assume-yes install puppet

and that worked fine, no errors.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2012-03-27T16:10:44.641578+00:00_

You shouldn't need to explicitly add puppet to the list of packages, as it is installed if you have a 'puppet' section in your cloud-config .

However, even with that, I can't reproduce failure on precise in an openstack cloud.

Could you add:
 output: {all: '| tee -a /var/log/cloud-init-output.log'}

to the cloud-config and then collect coud-init-output.log ?

The only thing I can guess is that networking isn't really up, and possibly the resolution of puppet.cihm is failing for you when puppet is installed.

On an unrelated note, with precise, there is a more straightforward way of feeding data to cloud images than using the OVF (but it does not work in natty).  See http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/view/head:/doc/nocloud/README for more info.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Rob Schmidt(rjeschmi)** wrote on 2012-03-27T17:58:17.331361+00:00_

Well that solved the problem :) I did notice my vm seemed to lack a default console. Maybe I've misconfigured that.

Retesting with output off to make sure it still fails.

Thanks for the link. It seems pretty similar except with the added meta-data (which is nice so I probably will use it).

puppet not configured to start, please edit /etc/default/puppet to enable
   ...done.
Setting up debconf-utils (1.5.42ubuntu1) ...
Setting up ruby (4.8) ...
Processing triggers for libc-bin ...
ldconfig deferred processing now taking place
ec2: 
ec2: #############################################################
ec2: -----BEGIN SSH HOST KEY FINGERPRINTS-----
ec2: 1024 a8:ed:78:b5:1a:fb:ce:94:62:e7:30:72:2c:bb:18:fe  root@havarti (DSA)
ec2: 256 30:7b:a7:a5:8a:24:16:3b:a5:a1:af:a2:2e:6e:27:5c  root@havarti (ECDSA)
ec2: 2048 bf:35:15:7d:76:cb:33:a6:01:c2:58:58:f3:d3:f1:d1  root@havarti (RSA)
ec2: -----END SSH HOST KEY FINGERPRINTS-----
ec2: #############################################################
cat: write error: Input/output error
cat: write error: Input/output error
cloud-init boot finished at Tue, 27 Mar 2012 17:51:41 +0000. Up 266.75 seconds

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Rob Schmidt(rjeschmi)** wrote on 2012-03-27T18:12:29.627510+00:00_

One time when I booted after having the console attached I got:

Err http://archive.ubuntu.com/ubuntu/ precise/main libreadline5 amd64 5.2-11
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main libruby1.8 amd64 1.8.7.352-2
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main ruby1.8 amd64 1.8.7.352-2
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main libruby all 4.8
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main libshadow-ruby1.8 amd64 1.4.1-8build1
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main augeas-lenses all 0.10.0-0ubuntu4
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main libaugeas0 amd64 0.10.0-0ubuntu4
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main libaugeas-ruby1.8 amd64 0.3.0-1.1ubuntu4
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main facter all 1.6.5-1ubuntu1
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main puppet-common all 2.7.10-1ubuntu1
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main puppet all 2.7.10-1ubuntu1
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main debconf-utils all 1.5.41ubuntu2
  Temporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ precise/main ruby all 4.8
  Temporary failure resolving 'archive.ubuntu.com'

which is consistent with not having networking up (probably solved by passing static IP through metadata service as you point out)


I can't tell if the problem is the console or not for sure, but I seemed to have 100% failure before configuring the console (maybe there was a console write that failed with nowhere to go, but how would we ever be able to tell? )...

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2012-03-27T19:47:30.940102+00:00_

Well, by default the console is going to the serial console in kvm.
If you're running kvm in graphical mode, I htink its ctrl-alt-3 to see it.  In curses mode its alt-3.

You can also invoke kvm with '-serial file:serial.log' and it will all go there.

Whats the kvm command you're running with ?

I'm not able to reproduce with:
 kvm -m 256  -curses -drive file=disk.img,if=virtio -drive file=seed.iso,if=virtio


I suspect what is happening is you're using guest networking and the network comes up to the guest, but the dns isn't working just yet.  'cloud-init start' (which would find either the OVF or the NoCloud datasources provided) should not start until after all /etc/network/interfaces configured interfaces are up (which would include 'eth0' as dhcp).

Is there something that needs to happen in your guest to get DNS working?

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Rob Schmidt(rjeschmi)** wrote on 2012-03-27T19:58:35.513088+00:00_

No, nothing particular for the DNS. Everything worked on the next run. I'm using dhcp to configure the network though. It eventually worked that time too (just delayed a bit).

I'm running it through libvirt, but I'll see if I can get you the command line that was failing.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Rob Schmidt(rjeschmi)** wrote on 2012-04-03T15:41:19.444297+00:00_

Ok, I think we can close this. I can't reproduce consistently with kvm (although it generally fails under libvirt).

If I properly configure the VM it works fine so no need to track this down. If anyone else reports a similar problem I'm happy to try and reproduce.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2012-08-16T13:48:29.734294+00:00_

marking wont fix per previous comment
